File: migrations/setup-genome.rs
import * as anchor from "@coral-xyz/anchor";
import { web3 } from "@coral-xyz/anchor";
import { getKeypairFromFile } from "@solana-developers/node-helpers";
import * as splToken from "@solana/spl-token";

async function main() {
  const payerPath = process.argv[2];
  const adminPath = process.argv[3];
  const mintPath = process.argv[4];

  if (!payerPath || !adminPath || !mintPath) {
    console.error("Usage: ts-node setup-genome.ts <payer.json> <admin.json> <mint.json>");
    process.exit(1);
  }

  const provider = anchor.AnchorProvider.env();
  anchor.setProvider(provider);

  const payer = await getKeypairFromFile(payerPath);
  const admin = await getKeypairFromFile(adminPath);
  const mintKeypair = await getKeypairFromFile(mintPath);

  console.log("Payer:", payer.publicKey.toBase58());
  console.log("Admin:", admin.publicKey.toBase58());
  console.log("Mint keypair:", mintKeypair.publicKey.toBase58());

  const mint = await splToken.createMint(
    provider.connection,
    payer,
    admin.publicKey,
    null,
    6,
    mintKeypair,
    null,
    splToken.TOKEN_2022_PROGRAM_ID
  );
  console.log("Genome Mint:", mint.toBase58());

  const adminAta = await splToken.createAssociatedTokenAccount(
    provider.connection,
    payer,
    mint,
    admin.publicKey,
    null,
    splToken.TOKEN_2022_PROGRAM_ID,
    splToken.ASSOCIATED_TOKEN_PROGRAM_ID
  );
  console.log("Admin ATA:", adminAta.toBase58());

  const mintTx = await splToken.mintTo(
    provider.connection,
    payer,
    mint,
    adminAta,
    admin,
    1000000000000,
    [],
    {},
    splToken.TOKEN_2022_PROGRAM_ID
  );
  console.log("Mint to admin tx:", mintTx);
}

main()
  .then(() => process.exit(0))
  .catch((error) => {
    console.error(error);
    process.exit(1);
  });

File: migrations/setup_genome.ts
import * as anchor from "@coral-xyz/anchor";
import { airdrop } from "../tests/utils";

async function main(): Promise<void> {
  const [
    deployerAddress,
    verifier1Address,
    verifier2Address,
    verifier3Address,
    operatorAddress,
    adminAddress,
    organizerAddress
  ] = process.argv.slice(2);

  const operator = new anchor.web3.PublicKey(operatorAddress);
  const admin = new anchor.web3.PublicKey(adminAddress);
  const deployer = new anchor.web3.PublicKey(deployerAddress);
  const organizer = new anchor.web3.PublicKey(organizerAddress);
  const verifier1 = new anchor.web3.PublicKey(verifier1Address);
  const verifier2 = new anchor.web3.PublicKey(verifier2Address);
  const verifier3 = new anchor.web3.PublicKey(verifier3Address);

  const recipients = [operator, admin, deployer, organizer, verifier1, verifier2, verifier3];

  await Promise.all(
    recipients.map(
      async (keypair) => await airdrop(keypair, 10)
    )
  );
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});
